### PR TITLE
Move rudder arming up into RC_Channel library

### PR DESCRIPTION
**Pull request by peterbarker**
This will make rudder-arming consistent across our vehicles.

One known behaviour change is that the rover arm interval has moved from 2 to 3 seconds (to be like Copter and Plane).

Other known changes in behaviour:
 - Copter and Rover will try to arm if your throttle is non-zero, and you will get a warning that your throttle is non-zero.  In master we don't try to arm.  I think this new behaviour is far preferable from a user's standpoint as they're *told* why they can't arm right now in the case they meant to arm!  The new code has the same behaviour as Plane in master

~~We can't move the rudder-specific arming checks into the RC_Channels library as ToyMode also arms with a reason of "rudder" and presumably wants the same checks~~  We did some surgery to fix this

### CODE REVIEW

**Comment by peterbarker** [comment]
Tested in SITL - standard arming and RUDDER_DISARM values.

File: ArduCopter/Copter.h [hunk 1]
```diff
@@ -820,10 +816,11 @@ class Copter : public AP_Vehicle {
     void motor_test_stop();
 
     // motors.cpp
-    void arm_motors_check();
     void auto_disarm_check();
     void motors_output();
     void lost_vehicle_check();
+    uint32_t auto_disarm_begin;
```
Comment (rmackay9) [inline ArduCopter/Copter.h:822]: normally we don't put the members with the method declarations.. there should be a place lower down where it can fit in.

File: ArduCopter/RC_Channel.cpp [hunk 2]
```diff
@@ -203,7 +203,7 @@ bool RC_Channel_Copter::do_aux_function(const aux_func_t ch_option, const AuxSwi
             if ((ch_flag == AuxSwitchPos::HIGH) &&
                 (copter.flightmode->allows_save_trim()) &&
                 (copter.channel_throttle->get_control_in() == 0)) {
-                copter.save_trim();
+                copter.g2.rc_channels.save_trim();
```
Comment (rmackay9) [inline ArduCopter/RC_Channel.cpp:206]: is it a bit weird this feature goes all the way down into the rc_channels object instead of calling the copter feature directly?

File: libraries/RC_Channel/RC_Channels.cpp [hunk 3]
```diff
@@ -234,6 +240,48 @@ uint32_t RC_Channels::enabled_protocols() const
...
+    const uint32_t now = AP_HAL::millis();
+    if (rudder_arm_timer == 0) {
+        // first time we've seen the attempt
+        rudder_arm_timer = now;
+        return;
+    }
+
+    if (now - rudder_arm_timer >= 3000) {
+        // time to try to arm or disarm:
+        if (channel->get_control_in() > 4000) {
```
Comment (tridge) [inline libraries/RC_Channel/RC_Channels.cpp:271]: why not reset timer when we call this? this looks like it calls continuously

**Review by tridge — CHANGES_REQUESTED** [review]

File: ArduCopter/RC_Channel.cpp (hunk 2 above)
Comment (peterbarker) [reply ArduCopter/RC_Channel.cpp:206]: There's nothing in these functions which make them Copter-specific - if we were insane we could let quadplane use the same functions in qstab!  Same with Sub if they used RC.  Less so with Rover :-)

Moving it up would mean save_trim was available on `rc()` which would make this a little less odd (and it would be in the parent class...)

Oh, and I forgot about Blimp, which could share the same functions.

File: ArduCopter/Copter.h (hunk 1 above)
Comment (peterbarker) [reply ArduCopter/Copter.h:822]: I've moved it and added a comment

File: libraries/RC_Channel/RC_Channels.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/RC_Channel/RC_Channels.cpp:274]: Fixed.

Comment (tridge) [reply libraries/RC_Channel/RC_Channels.cpp:274]: doesn't that fix break the 10s check for copter?

File: ArduCopter/RC_Channel.cpp (hunk 2 above)
Comment (rmackay9) [reply ArduCopter/RC_Channel.cpp:206]: ok, txs for the explanation.. I've thought about it a bit more and I think you're right that it's fine here.

**Comment by tridge** [comment]
@rmackay9 also noted on the call that he's happy for the 10s trim feature to be removed if whoever removes it ensures docs are updated

**Comment by rmackay9** [comment]
Docs for autotrim are here: https://ardupilot.org/copter/docs/autotrim.html

**Comment by rmackay9** [comment]
@robustini,

how do you feel about us removing the ability to start "auto trim" by holding the sticks in the arming position for 10+ seconds?

I think very few people use auto trim any more and even fewer know that it can be started using this arming method.. but what do you think?

File: libraries/RC_Channel/RC_Channels.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/RC_Channel/RC_Channels.cpp:270]: Yes, it did.  Fixed and tested.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 4]
```diff
@@ -1167,6 +1167,30 @@ bool AP_Arming::pre_arm_checks(bool report)
...
+
+        // only permit arming if the vehicle isn't being commanded to
+        // move via RC input
+        const RCMapper * rcmap = AP::rcmap();
+        if (rcmap == nullptr) {
+            check_failed(true, "No rcmap");
+            return false;
+        }
+        RC_Channel *c = rc().channel(rcmap->throttle() - 1);
+        if (c == nullptr) {
```
Comment (WickedShell) [inline libraries/AP_Arming/AP_Arming.cpp:1184]: Can we collapse `No rcmap` to be the same check failed as `No throttle channel`? It's still very accurate, it's a firmware specific thing, and it saves us the error string in all cases....

File: libraries/AP_Arming/AP_Arming.cpp [hunk 5]
```diff
@@ -1167,6 +1167,30 @@ bool AP_Arming::pre_arm_checks(bool report)
 
 bool AP_Arming::arm_checks(AP_Arming::Method method)
 {
+    if (method == AP_Arming::Method::RUDDER) {
+        if (get_rudder_arming_type() == AP_Arming::RudderArming::IS_DISABLED) {
```
Comment (WickedShell) [inline libraries/AP_Arming/AP_Arming.cpp:1171]: How do we have a code path where we issue the arm, but the type is disabled? That seems like something we'd want a pretty strong internal error on, if not an assert....

File: libraries/AP_Arming/AP_Arming.cpp [hunk 6]
```diff
@@ -1244,6 +1268,13 @@ bool AP_Arming::disarm(const AP_Arming::Method method, bool do_disarm_checks)
     if (!armed) { // already disarmed
         return false;
     }
+    if (method == AP_Arming::Method::RUDDER) {
+        // option must be enabled:
+        if (get_rudder_arming_type() != AP_Arming::RudderArming::ARMDISARM) {
```
Comment (WickedShell) [inline libraries/AP_Arming/AP_Arming.cpp:1273]: Same comment as above.

File: libraries/RC_Channel/RC_Channels.cpp [hunk 7]
```diff
@@ -234,6 +240,52 @@ uint32_t RC_Channels::enabled_protocols() const
...
+    }
+
+    const uint32_t now = AP_HAL::millis();
+    if (rudder_arm_timer == 0) {
+        // first time we've seen the attempt
+        rudder_arm_timer = now;
+    } else if (now - rudder_arm_timer >= 3000) {
+        rudder_arm_timer = 0;  // do not reattempt for a while
+        // time to try to arm or disarm:
+        if (channel->get_control_in() > 4000) {
```
Comment (WickedShell) [inline libraries/RC_Channel/RC_Channels.cpp:270]: It's fair to not do it now, but for future reference we shouldn't be hardcoding 4000 here. We should be fetching the channel's range, then comparing to a given percentage of that range. (That way we have an option of normalizing to something logical like a float or higher resolution in the future)

File: libraries/AP_Arming/AP_Arming.cpp (hunk 5 above)
Comment (peterbarker) [reply libraries/AP_Arming/AP_Arming.cpp:1343]: This helps centralise the arming checks.  This means we don't need to check if the type of arming is disabled in the caller - we've got the type being attempted already being passed in here for a reason...

**Comment by peterbarker** [comment]
I've put this onto a Rover and made sure it arms/disarms and checks the throttle.

I've put this onto a Plane and made sure it arms/disarms, honours RUDDER_ARMING and the throttle checks.

**Comment by peterbarker** [comment]
This needs testing on a Copter.

File: libraries/RC_Channel/RC_Channels.cpp [hunk 8]
```diff
@@ -341,6 +347,49 @@ RC_Channel &RC_Channels::get_yaw_channel()
...
+        // first time we've seen the attempt
+        rudder_arm_timer = now;
+        return;
+    }
+
+    if (now - rudder_arm_timer < 3000) {
+        // not time yet....
+        return;
+    }
+
```
Comment (IamPete1) [inline libraries/RC_Channel/RC_Channels.cpp:383]: I would have expected a `rudder_arm_timer = 0` here somewhere so you get one arm attempt per 3 seconds rather than continual attempts after 3 seconds.

Comment (peterbarker) [reply libraries/RC_Channel/RC_Channels.cpp:383]: Fixed, thanks.  It was a problem.

File: ArduPlane/radio.cpp [hunk 9]
```diff
@@ -116,60 +116,6 @@ void Plane::init_rc_out_aux()
...
-
-				if (rudder_arm_timer == 0) {
-                    rudder_arm_timer = now;
-                }
-			} else {
-				//time to arm!
-				arming.arm(AP_Arming::Method::RUDDER);
-                rudder_arm_timer = 0;
-                seen_neutral_rudder = false;
-                takeoff_state.rudder_takeoff_warn_ms = now;
```
Comment (IamPete1) [inline ArduPlane/radio.cpp:145]: This is lost I think. May not be a issue, just means we get the first "Takeoff waiting for rudder release" immediately instead of after 3 seconds.

**Comment by peterbarker** [comment]
No problems with i-term build up taking off in stab or loiter

copter auto no regression - spins as commanded....

Played with various Plane frames and all seems fine

File: ArduPlane/radio.cpp (hunk 9 above)
Comment (peterbarker) [reply ArduPlane/radio.cpp:145]: Fixed

File: libraries/AP_Arming/AP_Arming.cpp [hunk 10]
```diff
@@ -1697,6 +1697,21 @@ bool AP_Arming::pre_arm_checks(bool report)
 bool AP_Arming::arm_checks(AP_Arming::Method method)
 {
 #if AP_RC_CHANNEL_ENABLED
+    if (method == AP_Arming::Method::RUDDER) {
```
Comment (tridge) [inline libraries/AP_Arming/AP_Arming.cpp:1700]: if ARMING_CHECK=0 then will that mean arming/disarming by rudder input will always be enabled? arm_checks() not called?

Comment (peterbarker) [reply libraries/AP_Arming/AP_Arming.cpp:1700]: Fixed and test added.

File: Rover/AP_Arming_Rover.cpp [hunk 11]
```diff
@@ -146,6 +154,14 @@ bool AP_Arming_Rover::arm(AP_Arming::Method method, const bool do_arming_checks)
  */
 bool AP_Arming_Rover::disarm(const AP_Arming::Method method, bool do_disarm_checks)
 {
+    if (method == AP_Arming::Method::RUDDER) {
+        if (rover.g2.motors.active()) {
```
Comment (tridge) [inline Rover/AP_Arming_Rover.cpp:158]: this seems to be missing the throttle check that was in the rover code that did rudder input for arming

Comment (tridge) [reply Rover/AP_Arming_Rover.cpp:158]: motors active != check for neutral throttle

File: Rover/radio.cpp [hunk 12]
```diff
@@ -61,71 +61,6 @@ void Rover::init_rc_in()
...
-{
-    // check if arming/disarm using rudder is allowed
-    const AP_Arming::RudderArming arming_rudder = arming.get_rudder_arming_type();
-    if (arming_rudder == AP_Arming::RudderArming::IS_DISABLED) {
-        return;
-    }
-
-    // In Rover we need to check that its set to the throttle trim and within the DZ
-    // if throttle is not within trim dz, then pilot cannot rudder arm/disarm
-    if (!channel_throttle->in_trim_dz()) {
```
Comment (tridge) [inline Rover/radio.cpp:77]: we've lost this check in rover

**Review by tridge — CHANGES_REQUESTED** [review]
i think the big question is if when a user has non-zero throttle, and does full rudder deflection, is that considered an attempt to arm/disarm? ie. even if it fails in the arming checks, do we show them a message saying it failed, potentially worrying the user when they deliberately had the throttle non-zero

File: ArduCopter/motors.cpp [hunk 13]
```diff
@@ -6,73 +6,6 @@
...
-
-#if TOY_MODE_ENABLED
-    if (g2.toy_mode.enabled()) {
-        // not armed with sticks in toy mode
-        return;
-    }
-#endif
-
-    // ensure throttle is down
-    if (channel_throttle->get_control_in() > 0) {
```
Comment (tridge) [inline ArduCopter/motors.cpp:30]: this check is lost I think, in master we won't get a failed arm event, in this PR I think we will

File: Rover/radio.cpp (hunk 12 above)
Comment (peterbarker) [reply Rover/radio.cpp:77]: We have a control-in check instead.

File: ArduCopter/motors.cpp (hunk 13 above)
Comment (peterbarker) [reply ArduCopter/motors.cpp:30]: Yep, as we do in Plane, so we become consistent.  Something to discuss!

**Comment by peterbarker** [comment]
Flew it on SkyViper to test the ToyMode changes (you can't actually arm SkyViper via rudder with the standard controller due to the circular-angle problem!)

**Comment by peterbarker** [comment]
> ```
> Should we warn the user if they go to arm and their throttle is non-zero
> no
> Move gesture check into RC channels library
> Do not send message if user accidentally tries to arm vehicle with non-zero throttle value
> Custom changes to toymode
> Add enum value for toymode
> ```

All done

File: ArduCopter/AP_Arming_Copter.cpp [hunk 14]
```diff
@@ -566,6 +566,16 @@ bool AP_Arming_Copter::alt_checks(bool display_failure)
 //  has side-effect that logging is started
 bool AP_Arming_Copter::arm_checks(AP_Arming::Method method)
 {
+    if (method == AP_Arming::Method::RUDDER) {
```
Comment (rmackay9) [inline ArduCopter/AP_Arming_Copter.cpp]: as discussed, it should be safe to remove this bit of code because it shouldn't be a problem if users use the regular rudder arming gesture while in toymode

File: ArduPlane/AP_Arming_Plane.cpp [hunk 15]
```diff
@@ -254,24 +254,6 @@ bool AP_Arming_Plane::ins_checks(bool display_failure)
...
-            //parameter disallows rudder arming/disabling
-
-            // if we emit a message here then someone doing surface
-            // checks may be bothered by the message being emitted.
-            // check_failed(true, "Rudder arming disabled");
-            return false;
-        }
-
-        // if throttle is not down, then pilot cannot rudder arm/disarm
-        if (!is_zero(plane.get_throttle_input())){
```
Comment (tridge) [inline ArduPlane/AP_Arming_Plane.cpp:270]: this looks like a change, no longer honouring the reversed_throttle flag, which is set with an AUX function
I'm not sure this matters, but needs some thought

Comment (tridge) [reply ArduPlane/AP_Arming_Plane.cpp:270]: ahh, we're looking for zero, and -0 == 0, so OK?

File: ArduCopter/AP_Arming_Copter.cpp (hunk 14 above)
Comment (peterbarker) [reply ArduCopter/AP_Arming_Copter.cpp]: Removed it!

**Review by rmackay9 — APPROVED** [review]
I'm happy with this if we can remove the check for unused RC channels for rover and boat

**Review by tridge — APPROVED** [review]

File: Rover/AP_Arming_Rover.cpp (hunk 11 above)
Comment (andyp1per) [reply Rover/AP_Arming_Rover.cpp:166]: Did this get resolved?

File: libraries/RC_Channel/RC_Channels.cpp [hunk 16]
```diff
@@ -345,6 +351,71 @@ RC_Channel &RC_Channels::get_lateral_channel() const
...
+        rudder_arm_timer = 0;
+        return;
+    }
+
+    // enforce correct stick gesture for arming (but not disarming):
+    if (arming_check_throttle() && control_in > 4000) {
+        // only permit arming if the vehicle isn't being commanded to
+        // move via RC input
+        const auto &c = rc().get_throttle_channel();
+        if (c.get_control_in() != 0) {
```
Comment (andyp1per) [inline libraries/RC_Channel/RC_Channels.cpp:389]: don't you want to use the deadzone version here?

**Review by andyp1per — COMMENTED** [review]
Think its ok, just a couple of questions

File: ArduCopter/AP_Arming_Copter.cpp [hunk 17]
```diff
@@ -801,6 +801,12 @@ bool AP_Arming_Copter::disarm(const AP_Arming::Method method, bool do_disarm_che
         return false;
     }
 
+    if (method == AP_Arming::Method::RUDDER) {
+        if (!copter.flightmode->has_manual_throttle() && !copter.ap.land_complete) {
```
Comment (andyp1per) [inline ArduCopter/AP_Arming_Copter.cpp:805]: I'm wondering whether we should remove the manual throttle check here. Disarming mid-air at zero throttle, full yaw seems bad.

File: libraries/RC_Channel/RC_Channels.cpp (hunk 16 above)
Comment (peterbarker) [reply libraries/RC_Channel/RC_Channels.cpp:389]: I don't think so - this is pilot stick input we want to be inactive.  It would be really, really difficult for a Rover user to get zero input on a mid-stick-is-zero throttle input without a dz :-)

File: ArduCopter/AP_Arming_Copter.cpp (hunk 17 above)
Comment (rmackay9) [reply ArduCopter/AP_Arming_Copter.cpp:805]: Thanks AndyP very much for helping test this important PR

we should certainly be careful about changes in behaviour because it is also bad to be unable to disarm the vehicle waiting for the land-complete to become true.  That can lead to vehicles flopping around on the ground.

Comment (peterbarker) [reply ArduCopter/AP_Arming_Copter.cpp:805]: I haven't given this enough thought one way or the other.  I absolutely don't want to change something so fundamental/dangerous in this PR.

I *do* agree mid-air-accidental-disarms sound very bad.  But we might end up with people not able to disarm depending on how the landing detector modifies its behaviour based on manual throttle etc etc.

We *could* document our way out of it - suggest people doing inverted yaw spins may want to set the disarm-via-stick to false.  We could potentially change the default value to disallow it in minimize-fpv-osd setups.  We could say that if you've armed-with-a-switch-with-airmode then we don't do it.  Lots of options, lots of thought, a very small amount of coding, a lot of testing!

I'd also say that I've never actually heard of someone managing to hit this problem - I do find that a bit of a surprise.

tldr; sure, we need to think about it, but I do not want to do it as part of this PR.

File: Rover/AP_Arming_Rover.cpp (hunk 11 above)
Comment (peterbarker) [reply Rover/AP_Arming_Rover.cpp:166]: Yes, we have neutral-throttle checks.

File: ArduCopter/AP_Arming_Copter.cpp (hunk 17 above)
Comment (andyp1per) [reply ArduCopter/AP_Arming_Copter.cpp:805]: I was thinking that perhaps the right check was manual throttle without airmode, that would probably cover the majority case in a safe way.

Comment (peterbarker) [reply ArduCopter/AP_Arming_Copter.cpp:805]: > I was thinking that perhaps the right check was manual throttle without airmode, that would probably cover the majority case in a safe way.

Probably right - but not for this PR.

Are you OK with this one, @andyp1per ?
